js/ai/src/generate/action.ts

```diff
@@ -79,8 +81,20 @@ export const GenerateUtilParamSchema = z.object({
 export async function generateHelper(
   registry: Registry,
   input: z.infer<typeof GenerateUtilParamSchema>,
-  middleware?: ModelMiddleware[]
+  middleware?: ModelMiddleware[],
+  maxTurns?: number
 ): Promise<GenerateResponseData> {
+  maxTurns = maxTurns ?? 0;
+  const maxIterations = input.maxTurns ?? 5;
+  if (maxTurns > maxIterations) {
+    throw new GenkitError({
+      message: `Exceeded maximum tool call iterations (${maxIterations})`,
+      status: 'ABORTED',
+      detail: {
+        request: input,
```

Comment:
This should be a GenerationResponseError, I think, and include a response that includes all the stuff that was done until it bailed out.

Reply:
didn't know GenerationResponseError was a thing. :) Done.